xgcm/grid_ufunc.py

```diff
@@ -104,23 +100,84 @@ def _parse_grid_ufunc_signature(
     return in_ax_names, out_ax_names, in_ax_pos, out_ax_pos
 
 
+def _build_signature_from_type_hints(hints: Dict[str, Any]) -> str:
+
+    if "return" in hints:
+        return_hint = hints.pop("return")
+
+        # if ufunc returns multiple values (each of which might be annotated) we must extract from Tuple first
+        return_hints = (
+            list(return_hint.__args__)
+            if return_hint._name == "Tuple"
+            else [return_hint]
+        )
+
+        return_annotations = [
+            hint.__metadata__[0]
+            for hint in return_hints
+            if hasattr(hint, "__metadata__")
+        ]
+
+        return_arg_signature = ",".join(
+            [f"({variable_annotation})" for variable_annotation in return_annotations]
+        )
+    else:
+        return_arg_signature = "()"
+
+    arg_annotations = {
+        argname: hint.__metadata__[0]
+        for argname, hint in hints.items()
+        if hasattr(hint, "__metadata__")
+    }
+
+    # TODO check number of annotations?
+
+    if len(arg_annotations) > 0:
+        input_arg_signature = ",".join(
+            [
+                f"({variable_annotation})"
+                for variable_annotation in arg_annotations.values()
+            ]
+        )
+    else:
+        input_arg_signature = "()"
+
+    return f"{input_arg_signature}->{return_arg_signature}"
+
+
 class Signature:
     signature: str
-    in_ax_names: List[Tuple[str, ...]]
-    out_ax_names: List[Tuple[str, ...]]
-    in_ax_positions: List[Tuple[str, ...]]
-    out_ax_positions: List[Tuple[str, ...]]
+    in_ax_names: T_AX_POS_LIST
+    out_ax_names: T_AX_POS_LIST
+    in_ax_positions: T_AX_POS_LIST
+    out_ax_positions: T_AX_POS_LIST
 
     _REPLACEMENT_DUMMY_INDEX_NAMES = [f"__{char}" for char in string.ascii_letters]
 
     def __init__(self, signature: str):
-        self.signature: str = signature
+        self._parse(signature)
```

Comment:
Implementing the above suggestion would basically involve refactoring this class to allow a more direct construction path, e.g. 

```python
    def __init__(self, in_ax_names, out_ax_names, in_ax_positions, out_ax_positions)
```

and then adding another class method for `from_signature_string`.

Reply:
I've done this so I'm just going to merge this PR.